### PR TITLE
Inserting only non-duplicate chunks in MV

### PR DESCRIPTION
**Pull request by priera**
### Changelog category (leave one):
- Bug Fix

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed error that prevented insertion in replicated MV of new data in presence of duplicated data

### CODE REVIEW

**Comment by priera** [comment]
Technical description, feel free to comment or make any follow-up question.

This PR aims to fix the following case: when data is inserted in a replicated table, it may not be inserted in an associated MV if some of it (but not all) is duplicated.

As a quick example, if we have a table `landing`, with a target MV `mv`, the following statements won't insert the record `('2021-09-01 11:00:00', 24)` in `mv`. Please refer to the new test for the complete setup

```
INSERT INTO landing VALUES ('2020-01-01 13:23:34', 24);
INSERT INTO landing VALUES ('2021-09-01 11:00:00', 24), ('2020-01-01 13:23:34', 24);
```

Currently neither the case `Last block is duplicate` nor `One block is duplicate (changed setting)` pass in master. In both of them the new data is inserted in the table but not in the MV, since some of the blocks are duplicated.

The approach followed in order to solve this has been to change the behavior of the class `ReplicatedMergeTreeSinkImpl`:  the variable `last_block_is_duplicate` could prevent data from moving to the next stages of the query pipeline. Instead of using this variable, the class now builds a new chunk with just the non-duplicated data that will be forwarded to the next stages of the query graph. 

As a bit of simplifications due to these changes, I removed the method `lastBlockIsDuplicate` since doesn't seem to be necessary now. The changes applied in `SinkToStorage` seem to only directly affect the derived class `ReplicatedMergeTreeSinkImpl`

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 2b182fe3193f9a311c7865938430ca7c86de7140 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54184/2b182fe3193f9a311c7865938430ca7c86de7140/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by den-crane** [comment]
Interesting, probably it was the cause of lot of issues with the inconsistency between MatView and source table.

**Comment by priera** [comment]
Thanks to the test `01825_type_json_2` I noticed that my solution didn't work for tables with object columns. I'm working on a slightly different approach, and will push it early next week

**Comment by priera** [comment]
After checking, just one test failed `test_backup_restore_on_cluster/test.py::test_stop_other_host_during_backup[True]`, within the `Integration tests (asan) [5/6]` suite.

Same test ran successfully in other suites: asan-analyzer, release, tsan... I checked the logs and the errors don't seem related to sanitizing at all, instead there are exceptions of the kind: `DB::Exception: Backup Disk('backups', '1') is being written already. (BACKUP_ALREADY_EXISTS)` I think is possible this was a momentary error related to the environment where the test was running

File: src/Processors/Sinks/OutputChunkGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    Chunk in_chunk;
+    MutableColumns out_cols;
+    size_t row_offset = 0;
+    size_t final_chunk_rows = 0;
+};
+
+std::unique_ptr<IOutputChunkGenerator> IOutputChunkGenerator::createCopyRanges(ContextPtr context)
+{
+    // If MV is responsible for deduplication, block is not considered duplicated.
+    if (context->getSettingsRef().deduplicate_blocks_in_dependent_materialized_views)
```
Comment (hanfei1991) [inline src/Processors/Sinks/OutputChunkGenerator.cpp:72]: How about checking this outside? so that we do not have to pass `context`

**Review by hanfei1991 — APPROVED** [review]

**Pull request merged**